src/NummatDn03.jl

```diff
@@ -0,0 +1,273 @@
+    if t0==t1
+        return Dopri5Result([float(t0)],[Float64.(y0)])
+    end
+    dopri5!(f,t0,y0,t1;rtol=rtol,atol=atol,h_init=h_init)
+end
+
+#nihalo
+struct PendulumParams; g::Float64; l::Float64; end
+
+# right hand side za pendulum
+function pendulum_rhs(t,y,p::PendulumParams)
+    θ=y[1]; ω=y[2]
+    [ω, -(p.g/p.l)*sin(θ)]
+end
+
+
+
+"""
+theta_at_time(t; θ0, ω0, g=9.81, l=1.0)
+
+Vrne odmik nihala θ(t) pri času `t`.
+Parametri: začetni odmik `θ0` (rad), začetna hitrost `ω0` (rad/s), `g` in `l`.
+Uporablja lastno DOPRI5 integracijo.
+"""
+# θ(t)
+function theta_at_time(t; θ0,ω0,g=9.81,l=1.0,rtol=1e-9,atol=1e-12)
+    p=PendulumParams(float(g),float(l))
+    f=(τ,y)->pendulum_rhs(τ,y,p)
+    sol=solve_to(f,0.0,[float(θ0),float(ω0)],float(t);rtol=rtol,atol=atol)
+    sol.y[end][1]
+end
+
+energy(θ,ω;g=9.81,l=1.0)= g*l*(1-cos(θ))+0.5*(l^2)*ω^2
+
+
+
+
+"""
+harmonic_theta(t, θ0, ω0; g=9.81, l=1.0)
+
+Linearni (majhni odmiki) model nihala: θ(t) ≈ θ0 cos(√(g/l) t) + (ω0/√(g/l)) sin(√(g/l) t).
+Služi kot primerjava, ni “točen” za velike amplitude.
+"""
+function harmonic_theta(t,θ0,ω0;g=9.81,l=1.0)
+    wn=sqrt(g/l)
+    θ0*cos(wn*t)+(ω0/wn)*sin(wn*t)
+end
+
+advance_to(f,tL,yL,tR;rtol=1e-9,atol=1e-12)=
+    solve_to(f,tL,yL,tR;rtol=rtol,atol=atol).y[end]
+
+# refine zero crossing za θ=0
+function refine_zero_theta(f,tL,yL,tR,yR;rtol=1e-9,atol=1e-12,tol_time=1e-12)
+    θL=yL[1]; θR=yR[1]
+    while abs(tR-tL)>tol_time
+        tM=0.5*(tL+tR)
+        yM=advance_to(f,tL,yL,tM;rtol=rtol,atol=atol)
+        θM=yM[1]
+        if θL==0.0; return tL,yL
+        elseif θR==0.0; return tR,yR
+        end
+        if sign(θL)!=sign(θM); tR,yR,θR=tM,yM,θM
+        else; tL,yL,θL=tM,yM,θM; end
+    end
+    tZ=0.5*(tL+tR)
+    yZ=advance_to(f,tL,yL,tZ;rtol=rtol,atol=atol)
+    return tZ,yZ
+end
+
+
+
+"""
+find_period(θ0, ω0; g=9.81, l=1.0)
+
+Oceni periodo nihanja z iskanjem dveh zaporednih prehodov skozi θ=0 (smer navzgor, ω>0).
+Vrne čas v sekundah. Za večje amplitude je T > 2π√(l/g).
+"""
+# najdi periodo prek preseka θ=0, samo naraščajoč prehod
+function find_period(θ0,ω0;g=9.81,l=1.0,rtol=1e-9,atol=1e-12,maxT=1e4)
+    p=PendulumParams(float(g),float(l))
+    f=(τ,y)->pendulum_rhs(τ,y,p)
+
+    t=0.0; y=[float(θ0),float(ω0)]
+    h=0.05; crossings=Float64[]; last_t=t; last_y=copy(y)
+
+    while t<maxT && length(crossings)<2
+        sol=solve_to(f,t,y,t+h;rtol=rtol,atol=atol)
+        t=sol.t[end]; y=sol.y[end]
+
+        θL=last_y[1]; θR=y[1]
+        if (θL<0.0)&&(θR>0.0)
+            tZ,yZ=refine_zero_theta(f,last_t,last_y,t,y;rtol=rtol,atol=atol,tol_time=1e-12)
+            if yZ[2]>0.0; push!(crossings,tZ); end
+            t,y=tZ,yZ
+            sol2=solve_to(f,t,y,t+1e-6;rtol=rtol,atol=atol) # micro step
+            t=sol2.t[end]; y=sol2.y[end]
+        end
+        last_t=t; last_y=copy(y)
+        h=min(0.2,max(0.01,h))
+    end
+
+    if length(crossings)<2; error("no 2 crossings :("); end
+    crossings[end]-crossings[end-1]
```

Comment:
Natančnost izračuna periode je omejena s korakom Dopri5 metode. Ta bo bistveno prevelik, da zahtevano natančnostjo. Če korak preveč zmanjšate, pa bo čas za izračun vseh korakov bistveno narastel.

Iskanje periode formulirajte kot reševanje nelinearne enačbe in uporabite metode, ki so temu namenjene(Newtonova metoda, bisekcija, ...)